### PR TITLE
Mamba: add generative tests

### LINKED ISSUE #30828: Unable to run generation tests for Mamba & Jamba models
### System Info

- `transformers` version: 4.41.0.dev0
- Platform: Linux-5.15.0-1045-aws-x86_64-with-glibc2.31
- Python version: 3.10.9
- Huggingface_hub version: 0.23.0
- Safetensors version: 0.4.2
- Accelerate version: 0.29.2
- Accelerate config: 	not found
- PyTorch version (GPU?): 2.2.2+cu121 (True)
- Tensorflow version (GPU?): not installed (NA)
- Flax version (CPU?/GPU?/TPU?): 0.7.0 (cpu)
- Jax version: 0.4.13
- JaxLib version: 0.4.13
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help?

@gante @zucchini-nlp 

### Information

- [X] The official example scripts
- [ ] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [ ] My own task or dataset (give details below)

### Reproduction

See #30826

`test_assisted_decoding_matches_greedy_search_0_random` is forcibly skipped for Jamba because it's necessary to unset `_supports_cache_class` to resolve failing tests on main. 

`test_assisted_decoding_matches_greedy_search_0_random` appears to pass for Mamba, but this is because `all_generative_models` is not set in the model tester

### Expected behavior

Either `test_assisted_decoding_matches_greedy_search_0_random` can be run for both models with `_supports_cache_class` unset or it's not necessary to have `_supports_cache_class` unset for Jamba

### PR DESCRIPTION
**Pull request by gante**
Supercedes #31094

Fixes #30828 🤗

This PR:
1. Enables the generative tests to `mamba`
2. Removes redundant/needless manual skips through the `has_attentions` tester flag (TIL about this flag)
3. To enable (1.), adds a new flag to `PreTrainedModel` -- `_is_stateful` (self explanatory) [note: AFAIK we only have 3 stateful models, `mamba`, `jamba`, and `rwkv`, lmk if there are more!]
4. Uses `_is_stateful` to set a few appropriate exceptions and test skips

Note: with `has_attentions` and `_is_stateful`, I suspect I can remove a few more skips in a follow up PR :)

### CODE REVIEW

File: tests/models/mamba/test_modeling_mamba.py [hunk 1]
```diff
@@ -250,6 +250,8 @@ def prepare_config_and_inputs_for_common(self):
 @require_torch
 class MambaModelTest(ModelTesterMixin, GenerationTesterMixin, PipelineTesterMixin, unittest.TestCase):
     all_model_classes = (MambaModel, MambaForCausalLM) if is_torch_available() else ()
+    all_generative_model_classes = (MambaForCausalLM,) if is_torch_available() else ()
```
Comment (gante) [inline tests/models/mamba/test_modeling_mamba.py:253]: all tests passing for mamba :)

File: tests/models/mamba/test_modeling_mamba.py [hunk 2]
```diff
@@ -364,14 +362,6 @@ def test_initialization(self):
                         # check if it's a ones like
                         self.assertTrue(torch.allclose(param.data, torch.ones_like(param.data), atol=1e-5, rtol=1e-5))
 
-    @unittest.skip("Mamba does not use attention")
-    def test_attention_outputs(self):
-        r"""
-        Overriding the test_attention_outputs test as the attention outputs of Mamba are different from other models
-        it has a shape `batch_size, seq_len, hidden_size`.
-        """
-        pass
```
Comment (gante) [inline tests/models/mamba/test_modeling_mamba.py:373]: `has_attentions` skips this test

**Review by amyeroberts — APPROVED** [review]
Beautiful - thanks for adding. bonus points for skipping properly with skip messages ❤️

**Review by zucchini-nlp — APPROVED** [review]
Thanks for adding! Good to see some more usage of GenerationTesterMixin for non-standard models

**Pull request merged**